### PR TITLE
llama : preserve original batch order for speculative decoding layer inputs

### PR DESCRIPTION
**Pull request by hthadicherla**
## Overview

When llama.cpp splits a batch to process multiple requests (concurrency > 1), it can change the token order. With separate KV streams, the hidden states could reach DFlash in that changed order even though it expected the original order. This meant the draft sometimes received another token's hidden states, reducing acceptance.

The fix remembers each token's original position and puts its hidden states back in the correct row before DFlash reads them. This fixes the mismatch while keeping separate KV streams.

## Results

Tested on SPEED-Bench using Qwen3.8-27B with its DFlash2 drafter, both quantized to Q4_K_M with concurrency = 16.

Mean acceptance length (AL) improved from **3.3531 to 3.9211**, a **16.94% increase** overall.

| Category | Before fix AL | After fix AL | Improvement |
|---|---:|---:|---:|
| Coding | 3.4884 | 4.0518 | +16.15% |
| Humanities | 3.2335 | 3.6909 | +14.14% |
| Math | 3.7259 | 4.3573 | +16.94% |
| Multilingual | 3.6754 | 4.3538 | +18.46% |
| QA | 3.0643 | 3.4813 | +13.61% |
| RAG | 3.7012 | 4.4477 | +20.17% |
| Reasoning | 3.2405 | 3.7265 | +15.00% |
| Roleplay | 3.2166 | 3.6836 | +14.52% |
| STEM | 3.2221 | 3.7255 | +15.62% |
| Summarization | 3.1989 | 4.0703 | +27.24% |
| Writing | 3.1658 | 3.6926 | +16.64% |
| **Overall** | **3.3531** | **3.9211** | **+16.94%** |

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md).
- AI usage disclosure: Yes. I used AI to help write code and understand the codebase.

### CODE REVIEW

File: src/llama-context.cpp [hunk 1]
```diff
@@ -2233,18 +2234,29 @@ void llama_context::extract_layer_inputs(const llm_graph_result * res, size_t to
...
+            size_t end = i + 1;
+            while (end < n_tokens && batch_ids[end] == batch_ids[end - 1] + 1) {
+                ++end;
+            }
+            GGML_ASSERT(batch_ids[i] >= 0);
+            const size_t dst_offset = (size_t) batch_ids[i] * row_floats;
+            const size_t count = (end - i) * row_floats;
+            GGML_ASSERT(dst_offset + count <= embd_layer_inp[il].size);
+            ggml_backend_tensor_get_async(backend, t, embd_layer_inp[il].data + dst_offset,
+                    i * row_floats * sizeof(float), count * sizeof(float));
```
Comment (gaugarg-nv) [inline src/llama-context.cpp]: I think this is not compatible with `-sm tensor`. Check this: https://github.com/ggml-org/llama.cpp/blob/ebbb185227c31f1652f1445e2623563d2f67fe5a/ggml/src/ggml-backend-meta.cpp#L1914

Comment (hthadicherla) [reply src/llama-context.cpp]: Changed the copy to read each full tensor from the beginning, so it no longer uses the nonzero offsets that tensor mode rejects. Once the copy finishes, the CPU puts the hidden states back in the original token order before DFlash reads them. Modified test case to test for -sm tensor split mode as well.

Please review it again and see if the code looks okay now or not.

File: src/llama-context.cpp [hunk 2]
```diff
@@ -2214,7 +2215,10 @@ uint32_t llama_context::output_reserve(int32_t n_outputs) {
     return n_outputs_max;
 }
```
Comment (ruixiang63) [inline src/llama-context.cpp:2231]: This seems only fix DFlash/Eagle3, but not for MTP. I think MTP also has the same problem in `embd_nextn`, right?

Comment (hthadicherla) [reply src/llama-context.cpp:2231]: Just looked at the code. You are right that it does hit MTP as well.
Will make changes to extend support for MTP models as well.

Comment (hthadicherla) [reply src/llama-context.cpp:2231]: I’ve extended the fix to unmasked embd_nextn, using the original token order. Masked outputs keep the existing logits mapping. 

I also tested Qwen3.8-27B Q4_K_M with MTP on MT-Bench at concurrency 16. 
Mean AL improved from 2.5515 to 2.8027 (+9.85%).
throughput increased from 116.46 to 119.67 tokens/s in this run.


@ruixiang63 please review the code now and see if it looks okay or not.

**Comment by hthadicherla** [comment]
The latest commit addresses CI failures in the regression test I added. 
Debugging showed that WebGPU reserved too little scratch memory for smaller batches, while OpenVINO dropped requested hidden-state outputs and reused stale batch shapes and output configurations. 

The commit fixes these backend issues.

File: tests/test-llama-archs.cpp
Comment (ggerganov) [inline tests/test-llama-archs.cpp:1]: It's not clear what the added test is supposed to do. Better remove it from this PR and can be implemented properly in a follow-up.

**Review by ggerganov — COMMENTED** [review]
Drop the openvino and webgpu changes - these will be resolved by the respective maintainers later.

File: tests/test-llama-archs.cpp
Comment (hthadicherla) [reply tests/test-llama-archs.cpp:1]: Removed the test case and openvino and webgpu related changes.

Will properly implement and add the test case in future.

**Comment by hthadicherla** [comment]
I addressed the comments.
@ggerganov , can you please review it again ?

**Comment by ggerganov** [comment]
We'll have to coordinate this with the changes from #24669 .

I am wondering if there is some elegant way to write a test that confirms the order of the data is correct, so that we avoid such problems in the future. But it's not obvious how to do such test. LMK if you have some ideas.

**Comment by hthadicherla** [comment]
> I am wondering if there is some elegant way to write a test that confirms the order of the data is correct, so that we avoid such problems in the future. But it's not obvious how to do such test. LMK if you have some ideas.

I agree , my previous test was messy. 

I have an idea. What if we take two short, different sequences of unequal lengths. Decode each sequence independently, one token at a time, preserving its context. Copy each token’s hidden states into a table as reference. 

Then decode those same sequences concurrently together with seperate KV streams and small microbatch size to force splitting. We can compare the hidden states this provides with the reference and see if they come out to be same or not. 

The idea behind this test is that batching must not change which token owns a hidden-state row.

I will have to write the test and see if it works or not

> We'll have to coordinate this with the changes from https://github.com/ggml-org/llama.cpp/pull/24669 .

Should i rebase after that PR is merged ? It seems to change how a batch is constructed and how batch information is processed.
